src/db.c

```diff
@@ -943,9 +1007,15 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
         privdata[1] = o;
         do {
             cursor = dictScan(ht, cursor, scanCallback, NULL, privdata);
-        } while (cursor &&
-              maxiterations-- &&
-              listLength(keys) < (unsigned long)count);
+            /* In cluster mode there is a separate dictionary for each slot.
+             * If cursor is empty, we should try exploring next non-empty slot. */
+            if (o == NULL && !cursor) {
+                ht = dbGetNextUnvisitedSlot(c->db, &slot);
+            }
+        } while ((cursor || slot > 0) &&
```

Comment:
I believe the intention of `slot > 0` is for checking whether we are in cluster modes. Can we check that directly or use db->dict_count to do the check? 
We may hit a corner case by the current impl. Imagine we are scanning slot 0 in cluster mode, and we finished slot 0. Cursor would be 0 and slot was also 0, we quit. However, we shouldn't end here, we should continue with next slot.

Reply:
This check essentially tells you that there are more unvisited slots. Special case that you've described at slot 0 should work fine, as at the end of slot 0 iteration we'll find next non-empty slot, if there is one. There is also special case for the last slot, where -1 is returned, which stops the loop. This is why we have a check for 'slot>0' in 'addSlotIdToCursor', to not include slot id into cursor when iteration has finished. I can make it slightly more readable, but that would require a separate variable for holding "has more unvisited slots" status and would take couple more LoC. Would you prefer that, or does my explanation make sense?

Reply:
Perhaps I should also add a comment to make it more clear when reading the code.